### ISSUE TITLE
Day night cycle/RTC won't work in Pokémon Emerald Genesis 2.0 romhack

### ISSUE THREAD

**Issue by TU116** [mgba-emu/mgba#1721]
No matter what version or settings I use, my game is stuck in night mode in mGBA but works fine in VBA-M. I've heard most romhacks using custom RTC feature tend to have issues with non VBA emulators but I'd like to know if I'm missing something or if it's the fault of mGBA or the day night cycle implementation itself.

**Comment by RafaelTrepaUnCarballo** [comment]
Have you tried setting Realtime Clock on in Tools-->Game overrides...

**Comment by TU116** [comment]
Yes I've tried pretty much every settings related to RTC. Is game overrides supposed to be greyed out when opened while a game is already running ? Didn't do anything when I tried to force RTC anyway.

I forgot to mention I'm running W10 1809 x64 in case that's relevant. Tried v0.5, 0.8.1 and the latest dev build.

**Comment by endrift** [comment]
I look forward to that proof. Last I saw the day/night cycle code relied on VBA bugs.

**Comment by TU116** [comment]
[Link](https://www.pokecommunity.com/showthread.php?t=420288&page=8) to the thread, I ask the creator for further details.

**Comment by endrift** [comment]
Ok, since they're claiming it works on hardware I'll take a closer look before 0.9.

**Comment by endrift** [comment]
So the clock seems to work normally, but it's true that the game is stuck in night mode. I'm curious what they poll to get info about if it's day or night, since the clock itself seems fine.

**Comment by TriforceOfPwnage** [comment]
This seems to happen in Pokemon Blazing Emerald as well

**Comment by endrift** [comment]
It does not work on my EZFlash Omega, even though the clock does. Further, I've heard that similar Emerald ROM hacks don't work on Everdrive either in the same way. This is a problem with the hack's Day/Night System.

**Comment by TriforceOfPwnage** [comment]
i got it to work recently by re applying the dns with the modified dns.exe using your patch @endrift but that only applies to hacks using primedialga's rtc. i haven't been able to find a fix for shiny quagsires day and night system. i can post a patch file of shiny quagsires tho if you'd like to inspect it.

**Comment by TriforceOfPwnage** [comment]
Here's the asm files they used for reference 
[Shiny Quagsire Asm.zip](https://github.com/mgba-emu/mgba/files/5048406/Shiny.Quagsire.Asm.zip)

**Comment by endrift** [comment]
If @shinyquagsire23 were to fix that I wonder if people would use the new version instead.

**Comment by TriforceOfPwnage** [comment]
excuse my idiot brain. But do you mean fix the day night and seasons tool? Or Shinyquagsire's dns? the day night and seasons tool is made by primedialga. Which is what you fixed and was what got Emerald Genesis to work. I'm a bit sleep deprived so excuse my lack of comprehension rn.

**Comment by endrift** [comment]
I fixed the tool by editing the patch it applies. I'm not too familiar with the specific versions of the patches and patching tools.

**Comment by manuth** [comment]
Hi @endrift 
I stumbled accross this issue looking for a way to get the RTC for another ROM Hack to work in mGBA (https://github.com/WodkaRHR/Violet.git to be exact).

Is there any other way other than buying an EZFlash Omega Cartridge to test whether it works on real hardware?
Is running the ROM on a 3DS system using the `New Super Ultimate Injector` (https://3ds.eiphax.tech/nsui) a valid test?

**Comment by endrift** [comment]
I don't know if the 3DS VC supports the RTC. @profi200 would though.

Looking at the source it looks like it should work (if it's enabled--make sure to check the override view), but I'll have to take a closer look.

**Comment by profi200** [comment]
Yes, official GBA VC/Ambassador games can use the RTC. It's supported in hardware with the only caveat of being limited to the flash save types.

On a side note: Don't use open_agb_firm for RTC accuracy tests. I don't have the time and date offsets implemented yet. I just force the current date and time at every startup which i found breaks a few official games and a bunch of ROM hacks.

**Comment by manuth** [comment]
> I don't know if the 3DS VC supports the RTC. @profi200 would though.
> 
> Looking at the source it looks like it should work (if it's enabled--make sure to check the override view), but I'll have to take a closer look.

Awesome! 😄 
Thanks a ton for your effort.

I tried it on both mGBA (0.10 and 0.9.3) and my 3DS.
With the RTC override and VBA compatibility enabled, I receive `0`-values when trying to determine the time:
`Monday, 0-0-0, 0:0:0`

With RTC override enabled and VBA compatibility disabled (and also when running on my 3DS), I'm getting `165` for each date component:
` , 165-165-165, 165:165:165`

**Comment by manuth** [comment]
> Yes, official GBA VC/Ambassador games can use the RTC. It's supported in hardware with the only caveat of being limited to the flash save types.
> 
> On a side note: Don't use open_agb_firm for RTC accuracy tests. I don't have the time and date offsets implemented yet. I just force the current date and time at every startup which i found breaks a few official games and a bunch of ROM hacks.

Thank you so much 😄 
I just gave it a try using `open_agb_firm` and get the `165`-values once more.
Not sure as to why the RTC doesn't work on actual hardware.

**Comment by manuth** [comment]
@endrift sorry for pinging you once more
I just gave it a go to re-program the RTC portion and while doing so, I got a new log message:
```
Invalid RTC command byte: 63
```
And also
```
Invalid RTC command byte: 00
```

What's funny is, that I actually do send the command `0x63`. According to GBATEK's docs, this is the `time (hh,mm,ss)` command.

However, GBATEK tells me that it's recommended to always send commands LSB first rather than MSB first. However, if I were to send the command MSB first, I think, the check in this code-section should run through properly:
https://github.com/mgba-emu/mgba/blob/9adad40b1ce14b00a19a8544ba0742f9bf3342fb/src/gba/cart/gpio.c#L201

Question:
* Should commands - unlike pointed out in GBATEK - be sent MSB first?
* Does the actual hardware only allow MSB-first RTC commands as well?

Thank you so much for your help, I really appreciate your time 😄
